include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -189,68 +203,124 @@ struct BFloat16_t {
- * not owned by the caller)
+ * All of the C++ classes function as
+ *  a) smart pointers that own objects that are created by the underlying C API. Their size is just
+ *  a pointer size, no need to dynamically allocate them. Use them by value.
+ *  b) convenient interfaces that return C++ objects and further enhance exception and type safety.
  *
  */
+
+/// <summary>
+/// This is a non-const pointer holder that is move-only. Disposes of the pointer on destruction.
+/// </summary>
 template <typename T>
 struct Base {
+ public:
   using contained_type = T;
 
-  Base() = default;
-  Base(T* p) : p_{p} {
-    if (!p)
-      ORT_CXX_API_THROW("Allocation failure", ORT_FAIL);
-  }
+  constexpr Base() = default;
+  constexpr explicit Base(contained_type* p) noexcept : p_{p} {}
   ~Base() { OrtRelease(p_); }
 
-  operator T*() { return p_; }
-  operator const T*() const { return p_; }
-
-  /// \brief Releases ownership of the contained pointer
-  T* release() {
+  /// \brief Relinquishes ownership of the contained C object pointer
+  /// The underlying object is not destroyed
+  contained_type* release() {
     T* p = p_;
     p_ = nullptr;
     return p;
   }
 
- protected:
   Base(const Base&) = delete;
   Base& operator=(const Base&) = delete;
+
   Base(Base&& v) noexcept : p_{v.p_} { v.p_ = nullptr; }
   void operator=(Base&& v) noexcept {
     OrtRelease(p_);
     p_ = v.release();
   }
 
-  T* p_{};
+  constexpr operator contained_type*() const noexcept { return p_; }
 
-  template <typename>
-  friend struct Unowned;  // This friend line is needed to keep the centos C++ compiler from giving an error
+ protected:
+  contained_type* p_{};
 };
 
-/** \brief Wraps an object that inherits from Ort::Base and stops it from deleting the contained pointer on destruction
- *
- * This has the effect of making it not own the memory held by Ort::Base.
- */
+/// <summary>
+/// This is a const T template specialization of Base<T>. This accepts a const pointer
+/// to the underlying C object and does not destroy it, as we do not own it.
+/// </summary>
+/// <param name="p"></param>
 template <typename T>
-struct Unowned : T {
-  Unowned(typename T::contained_type* p) : T{p} {}
-  Unowned(Unowned&& v) : T{v.p_} {}
-  ~Unowned() { this->release(); }
+struct Base<const T> {
+ public:
+  using contained_type = const T;
+
+  constexpr Base() = default;
+  constexpr explicit Base(contained_type* p) noexcept : p_{p} {}
+
+  ~Base() = default;
+
+  // Allow copying of non-owning instances
+  Base(const Base&) = default;
+  Base& operator=(const Base&) = default;
+
+  Base(Base&& v) noexcept : p_{v.p_} { v.p_ = nullptr; }
+  void operator=(Base&& v) noexcept {
+    p_ = v.p_;
+    v.p_ = nullptr;
+  }
+
+  /// <summary>
+  /// XXX Should we allow assignmets of non-const Base to const Base
+  /// and non-const Unowned to const base?
+  /// </summary>
+
+  constexpr operator contained_type*() const noexcept { return p_; }
+
+ protected:
+  contained_type* p_{};
```

Comment:
Why is this being reverted back? This was removed in https://github.com/microsoft/onnxruntime/pull/8948

Reply:
The reverted was  a half baked solution which did not solve the problems at hand. I am going to update the PR motivation section for clarity.